play-validations/memory-footprint/src/main/java/com/google/wear/watchface/dfx/memory/AndroidManifest.kt

```diff
@@ -0,0 +1,161 @@
+package com.google.wear.watchface.dfx.memory
+
+import com.android.aapt.Resources.XmlNode
+import com.android.tools.apk.analyzer.BinaryXmlParser
+import com.android.tools.build.bundletool.commands.DumpManagerUtils
+import com.android.tools.build.bundletool.model.BundleModule
+import com.android.tools.build.bundletool.model.BundleModuleName
+import com.android.tools.build.bundletool.model.ZipPath
+import com.android.tools.build.bundletool.model.utils.xmlproto.XmlProtoNode
+import com.android.tools.build.bundletool.xml.XPathResolver
+import com.android.tools.build.bundletool.xml.XmlNamespaceContext
+import com.android.tools.build.bundletool.xml.XmlProtoToXmlConverter
+import org.w3c.dom.Document
+import java.io.ByteArrayInputStream
+import java.io.InputStream
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.zip.ZipEntry
+import java.util.zip.ZipFile
+import java.util.zip.ZipInputStream
+import javax.xml.namespace.NamespaceContext
+import javax.xml.parsers.DocumentBuilderFactory
+import javax.xml.xpath.XPathExpressionException
+import javax.xml.xpath.XPathFactory
+
+const val ANDROID_MANIFEST_FILE_NAME = "AndroidManifest.xml"
+
+/**
+ * Class that represents important properties of the AndroidManifest.xml file, for use in working
+ * with watch face packages.
+ */
+class AndroidManifest private constructor(
+    val wffVersion: Int,
+    val minSdkVersion: Int,
+    val targetSdkVersion: Int
+) {
+    companion object {
+        @JvmStatic
+        fun loadFromAab(zipFile: ZipFile): AndroidManifest {
+            val manifestPath =
+                ZipPath.create(BundleModuleName.BASE_MODULE_NAME.name)
+                    .resolve(BundleModule.SpecialModuleEntry.ANDROID_MANIFEST.path)
+            val manifestProto =
+                XmlProtoNode(
+                    DumpManagerUtils.extractAndParse(
+                        zipFile, manifestPath
+                    ) { input: InputStream? -> XmlNode.parseFrom(input) })
+
+            val doc = XmlProtoToXmlConverter.convert(manifestProto)
+
+            try {
+                val minSdk = getAttribute(
+                    doc,
+                    manifestProto,
+                    "//uses-sdk/@android:minSdkVersion"
+                ).toIntOrNull() ?: 1
+                val targetSdk = getAttribute(
+                    doc,
+                    manifestProto,
+                    "//uses-sdk/@android:targetSdkVersion"
+                ).toIntOrNull() ?: minSdk
+                val wffVersion =
+                    getAttribute(doc, manifestProto, "//property/@android:value").toInt()
+                return AndroidManifest(wffVersion, minSdk, targetSdk)
+            } catch (e: XPathExpressionException) {
+                throw RuntimeException(e)
+            }
```

Comment:
try-catch can be discarded, since this is Kotlin. This XPathExpressionException in particular can be ignored because the XPath strings are static, so you could only get them if the code is wrong.

Reply:
Good point! Thanks, I hadn't thought about this